File: java-to-go-projects/10-tcp-chat-server/Room.java
package chat;

import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * A chat room that holds a set of connected clients.
 * Supports broadcasting messages to all clients in the room.
 */
public class Room {
    private final String name;
    private final List<ClientHandler> clients = new CopyOnWriteArrayList<>();
    private final List<Message> history = new ArrayList<>();
    private static final int MAX_HISTORY = 50;

    public Room(String name) {
        this.name = name;
    }

    public String getName() { return name; }

    public void join(ClientHandler client) {
        clients.add(client);
        broadcast(new Message("SERVER", name,
                client.getUsername() + " joined the room"));
    }

    public void leave(ClientHandler client) {
        clients.remove(client);
        broadcast(new Message("SERVER", name,
                client.getUsername() + " left the room"));
    }

    /**
     * Send a message to all clients in this room.
     */
    public void broadcast(Message msg) {
        synchronized (history) {
            history.add(msg);
            if (history.size() > MAX_HISTORY) {
                history.remove(0);
            }
        }
        for (ClientHandler client : clients) {
            client.send(msg.format());
        }
    }

    public List<String> getMemberNames() {
        return clients.stream().map(ClientHandler::getUsername).toList();
    }

    public int size() { return clients.size(); }
}

File: java-to-go-projects/10-tcp-chat-server/rust/src/room.rs
use crate::message::Message;
use tokio::sync::broadcast;

pub struct Room {
    pub name: String,
    pub tx: broadcast::Sender<String>,
}

impl Room {
    pub fn new(name: &str) -> Self {
        let (tx, _) = broadcast::channel(100);
        Room {
            name: name.to_string(),
            tx,
        }
    }

    /// Broadcast a message to all clients in this room.
    pub fn broadcast(&self, msg: &Message) {
        todo!()
    }

    /// Get a new receiver for this room's broadcast channel.
    pub fn subscribe(&self) -> broadcast::Receiver<String> {
        self.tx.subscribe()
    }
}
